sound/soc/sof/intel/hda.c

```diff
@@ -499,6 +499,44 @@ static const struct sof_intel_dsp_desc
 	return chip_info;
 }
 
+irqreturn_t hda_dsp_interrupt_handler(int irq, void *context)
+{
+	struct snd_sof_dev *sdev = context;
+
+	if (is_ipc_irq(sdev) || is_stream_irq(sdev))
+		return IRQ_WAKE_THREAD;
+
+	return IRQ_NONE;
+}
+
+irqreturn_t hda_dsp_interrupt_threaded(int irq, void *context)
+{
+	struct snd_sof_dev *sdev = context;
+
+	if (is_ipc_irq(sdev))
+		schedule_work(&sdev->ipc_work);
+	if (is_stream_irq(sdev))
+		schedule_work(&sdev->stream_work);
```

Comment:
@RanderWang hda_dsp_stream_threaded_handler() could take a long time, and we don't wan't one threaded handler to wait for the other. Besides, we may want to use loop on each threaded handler in the future.

Reply:
@RanderWang has a point though.what would happen if we didn't not use a threaded handler, and called schedule_work directly from the handler?

Reply:
@bardliao I'm a bit concerned about this as well. We are essentially adding two mandatory context switches for every interrupt. First from hard-irq to irq thread, and then from irq thread to work-queue. Do you have concrete logs showing that the threaded handler is taking a lot of time?

Reply:
I also agree that this seems a bit of an overkill. Instead we should just wake the appropriate thread in the handler based on is_ipc_irq() or is_stream_irq() is true.

Reply:
Thank you all. Yes, it would be better to wake the appropriate thread in the handler.

Reply:
> @RanderWang hda_dsp_stream_threaded_handler() could take a long time, and we don't wan't one threaded handler to wait for the other. Besides, we may want to use loop on each threaded handler in the future.

Actually, the two works are still scheduled on the same CPU because scheduler is designed to utilize cache data (please check scheduler documents), so this can't save time and spend extra time to schedule work.

Reply:
> @RanderWang has a point though.what would happen if we didn't not use a threaded handler, and called schedule_work directly from the handler?

We should follow the requirement of request_threaded_irq
please check the comments for request_threaded_irq in kernel.
````
 *	If you want to set up a threaded irq handler for your device
 *	then you need to supply @handler and @thread_fn. @handler is
 *	still called in hard interrupt context and has to check
 *	whether the interrupt originates from the device. If yes it
 *	needs to disable the interrupt on the device and return
 *	IRQ_WAKE_THREAD which will wake up the handler thread and run
 *	@thread_fn. This split handler design is necessary to support
 *	shared interrupts.
````

Reply:
yes, this was a completely dumb comment :-)
I think what I was trying to say if whether the same handler can trigger multiple threads, but that seems like asking for trouble.